java/ql/src/semmle/code/java/security/XPath.qll

```diff
@@ -1,65 +1,56 @@
 /** Provides classes to reason about XPath vulnerabilities. */
 
 import java
-import semmle.code.java.dataflow.FlowSources
-import semmle.code.java.dataflow.TaintTracking
+import semmle.code.java.dataflow.DataFlow
+import semmle.code.java.dataflow.ExternalFlow
 
 /**
- * An abstract type representing a call to interpret XPath expressions.
+ * A sink that represents a method that interprets XPath expressions.
+ * Extend this class to add your own XPath Injection sinks.
  */
-class XPathSink extends MethodAccess {
-  /**
-   * Gets the argument representing the XPath expressions to be evaluated.
-   */
-  abstract Expr getSink();
-}
-
-/** The class `javax.xml.xpath.XPath` */
-class XPath extends RefType {
-  XPath() { this.hasQualifiedName("javax.xml.xpath", "XPath") }
-}
-
-/** A call to `XPath.evaluate` or `XPath.compile` */
-class XPathEvaluateOrCompile extends XPathSink {
-  XPathEvaluateOrCompile() {
-    exists(Method m | this.getMethod() = m and m.getDeclaringType() instanceof XPath |
-      m.hasName(["evaluate", "compile"])
-    )
+abstract class XPathInjectionSink extends DataFlow::Node { }
+
+/** CSV sink models representing methods susceptible to XPath Injection attacks. */
+private class DefaultXPathInjectionSinkModel extends SinkModelCsv {
+  override predicate row(string row) {
+    row =
+      [
+        "javax.xml.xpath;XPath;true;evaluate;;;Argument[0];xpath",
+        "javax.xml.xpath;XPath;true;evaluateExpression;;;Argument[0];xpath",
+        "javax.xml.xpath;XPath;true;compile;;;Argument[0];xpath",
+        "org.dom4j;Node;true;selectObject;;;Argument[0];xpath",
+        "org.dom4j;Node;true;selectNodes;;;Argument[0..1];xpath",
+        "org.dom4j;Node;true;selectSingleNode;;;Argument[0];xpath",
+        "org.dom4j;Node;true;numberValueOf;;;Argument[0];xpath",
+        "org.dom4j;Node;true;valueOf;;;Argument[0];xpath",
+        "org.dom4j;Node;true;matches;;;Argument[0];xpath",
+        "org.dom4j;Node;true;createXPath;;;Argument[0];xpath",
+        "org.dom4j;DocumentFactory;true;createPattern;;;Argument[0];xpath",
+        "org.dom4j;DocumentFactory;true;createXPath;;;Argument[0];xpath",
+        "org.dom4j;DocumentFactory;true;createXPathFilter;;;Argument[0];xpath",
+        "org.dom4j;DocumentHelper;false;createPattern;;;Argument[0];xpath",
+        "org.dom4j;DocumentHelper;false;createXPath;;;Argument[0];xpath",
+        "org.dom4j;DocumentHelper;false;createXPathFilter;;;Argument[0];xpath",
```

Comment:
DocumentHelper also has:
- [`selectNodes`](https://javadoc.io/static/org.dom4j/dom4j/2.1.3/org/dom4j/DocumentHelper.html#selectNodes(java.lang.String,java.util.List))
- [`sort`](https://javadoc.io/static/org.dom4j/dom4j/2.1.3/org/dom4j/DocumentHelper.html#sort(java.util.List,java.lang.String)) (not sure how relevant that is since it is apparently merely sorting nodes)

Reply:
Good catch! About `sort`, I think it's relevant - as I see it, depending on how the sorted nodes are used, it could enable a [blind XPath Injection attack](https://owasp.org/www-community/attacks/Blind_XPath_Injection).

I added these sinks in 0362e1c.